### DISCUSSION TITLE
Variable shadow/hilight sharpening

### DISCUSSION THREAD

**Discussion by ABLomas** [ImageMagick/ImageMagick#4015] [Digital Image Processing] [answered]
For final resizing (downsampling) i'm currently using software, called "ON1 Resize". It has very bad UI, i automated most of it using autohotkey, still not perfect - but it has one good feature - configurable sharpening for different regions (shadows/hilight). In other words - for final downsample can choose how to sharpen images - entire image (rare), hilights (for daytime landscapes....), shadows (cityscapes...).
I tried to do similar things with imagemagick using earlier discussions - combined https://legacy.imagemagick.org/discourse-server/viewtopic.php?t=33388 and http://im.snibgo.com/resamphm.htm in manual way - it kinda works (not entirely sure :D ) but produce sharp line between s/h split.
What is needed - "smooth masking", for example, when using script for shadows:
 - completely dark ares get 100% sharpening (not much to see, but anyway, all details sharpened)
 - dark zones get something like 75%
 - midtones 50-25%
 - hilights - 0%
In other words - no "sharp separation" between sharpened zones and just resampled.

Guess this can be "emulated" by using many separate masks, or there is better way?

**Comment by snibgo** [comment, marked as answer]
In the resampHM.bat script, the sharpness amount is a constant across the entire image. We can do the operation twice (or more) with different sharpness amounts, and combine these with masks.

If we use a sharp binary mask (eg "-threshold 50%"), this will usually cause visible problems at the boundary. This is because each output pixel is influenced by surrounding input pixels, so output pixels on either side of a sharp boundary will have clearly different influences.

The cure is to ensure the mask is not sharp. Instead of threshold, we can use "-level 40,60%" or similar. Or we can blur the mask. Or do both.

Windows BAT script:
```
call %PICTBAT%resampHM %SRC% 300x200 d 0 . . rhm_d_0.png

call %PICTBAT%resampHM %SRC% 300x200 d 100 . . rhm_d_100.png

%IMG7%magick ^
  rhm_d_0.png ^
  rhm_d_100.png ^
  -alpha off ^
  ( -clone 0 ^
    -colorspace Gray ^
    -level 40,60%% ^
    -blur 0x2 ^
    +write g2.png ^
  ) ^
  -compose Over -composite ^
  c2.png
```
We save the mask as g2.png just so we can see it:
![g2](https://user-images.githubusercontent.com/33812368/127818966-401e6c55-a8c7-4308-a861-946bfd8ff505.png)

The output c2.png is sharpened in the highlights, and not the shadows:
![c2](https://user-images.githubusercontent.com/33812368/127819006-a14dbc35-a1bd-456a-93e7-13bd6fa2519c.png)

EDIT to add: Of course, we don't need to partition the mask at all. No threshold or level or blur. Just leave it as continuous grayscale, so the gradation of sharpness is totally smooth.
```
%IMG7%magick ^
  rhm_d_0.png ^
  rhm_d_100.png ^
  -alpha off ^
  ( -clone 0 ^
    -colorspace Gray ^
    +write g3.png ^
  ) ^
  -compose Over -composite ^
  c3.png
```
![c3](https://user-images.githubusercontent.com/33812368/127820576-cbd5b94c-adcb-496d-a008-78be2dacc7b6.png)

For this image, the mask does have sharp boundaries, so a small blur would be helpful.

**Reply by ABLomas** [reply]
Thanks!
But some clarification needed about examples - so, what you did - called resample script, which did two copies of image - one with 0 sharpening, another one with 100% sharpening. Then you blend them both using mask.
Right?

**Reply by ABLomas** [reply]
Played a bit with "manual version" (echo on, all expanded to fixed variables list) - works good so far.
Some issues with noise resampling (when sharpening is set to hilights - noise (mostly visible on shadows) is usually oversampled (by downsizing 2x or even more) and not visible when using other software, but not with IM) - but i need to play more with variables and values. Using grayscale mask is good idea, no "sharp edges" between zones.
Will spend more time, should work well, thanks!

**Comment by snibgo** [comment]
> But some clarification needed about examples - so, what you did - called resample script, which did two copies of image - one with 0 sharpening, another one with 100% sharpening. Then you blend them both using mask. Right?

Exactly right.

For performance, a version of the script would call "magick" just once, and read each input just once, without creating intermediate results.

Sharpening will emphasise noise. Noise is best dealt with earlier in the process, ideally before demosaicing.

**Reply by ABLomas** [reply]
Played a bit with various settings, works good enough to be as replacement for "on1 resize".
I was unable to do entire transformation without temporary files (maybe doable, but i really don't know how to push temporary results between various processes on windows), but even with temp. files it's not bad (processing 100MB+ TIFF in ~6s).
Thanks!
